cli/cmd/generate.go

```diff
@@ -0,0 +1,348 @@
+		PreRunE: func(cmd *cobra.Command, _ []string) error {
+			// Load any cached inputs if interactive
+			if cli.InteractiveMode() {
+				cachedOptions := &aws.GenerateAwsTfConfigurationArgs{}
+				if ok := cli.ReadCachedAsset("iac-aws-generate-params", &cachedOptions); ok {
+					cli.Log.Debug("loaded previously set values for AWS iac generation")
+				}
+
+				extraState := &AwsGenerateCommandExtraState{}
+				if ok := cli.ReadCachedAsset("extra-state", &extraState); ok {
+					cli.Log.Debug("loaded previously set values for AWS iac generation (extra state)")
+				}
+
+				// Merge cached inputs to current options (current options win)
+				if err := mergo.Merge(GenerateAwsCommandState, cachedOptions); err != nil {
+					return errors.Wrap(err, "failed to load saved options!")
+				}
+				if err := mergo.Merge(GenerateAwsCommandExtraState, extraState); err != nil {
+					return errors.Wrap(err, "failed to load saved options!")
+				}
+			}
+
+			// Parse passed in subaccounts (if any)
+			if GenerateAwsCommandExtraState.AwsSubAccounts != "" {
+				// validate consolidated_cloudtrail is enabled - otherwise this flag cannot be used
+				if ok, _ := cmd.Flags().GetBool("consolidated_cloudtrail"); !ok {
+					return errors.New("aws subaccounts can only be supplied with consolidated cloudtrail enabled")
+				}
+
+				// validate the format of supplied value is correct
+				matchRegEx := `(\w+:\w+(,)?)+`
+				if ok, _ := regexp.MatchString(matchRegEx, GenerateAwsCommandExtraState.AwsSubAccounts); !ok {
+					return errors.New("supplied aws subaccounts in invalid format")
+				}
+
+				awsSubaccounts := []aws.AwsSubAccount{}
+				for _, account := range strings.Split(strings.TrimRight(GenerateAwsCommandExtraState.AwsSubAccounts, ","), ",") {
+					accountDetails := strings.Split(account, ":")
+					awsSubaccounts = append(awsSubaccounts, aws.NewAwsSubAccount(accountDetails[0], accountDetails[1]))
+				}
+
+				GenerateAwsCommandState.SubAccounts = awsSubaccounts
+			}
+
+			// Collect and/or confirm parameters
+			err := promptAwsGenerate(GenerateAwsCommandState, GenerateAwsExistingRoleState, GenerateAwsCommandExtraState)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+)
+
+func init() {
+	// add the iac-generate command
+	rootCmd.AddCommand(generateTfCommand)
+
+	// Add global flags for iac generation
+	generateTfCommand.PersistentFlags().StringVar(
+		&GenerateAwsCommandExtraState.Output, "output", "", "location to write generated content")
+
+	// add flags to sub commands
+	// TODO Share the help with the interactive generation
+	generateAwsTfCommand.PersistentFlags().BoolVar(
+		&GenerateAwsCommandState.Cloudtrail, "cloudtrail", false, "enable cloudtrail integration")
+	generateAwsTfCommand.PersistentFlags().BoolVar(
+		&GenerateAwsCommandState.Config, "config", false, "enable config integration")
+	generateAwsTfCommand.PersistentFlags().StringVar(
+		&GenerateAwsCommandState.AwsRegion, "aws_region", "", "specify aws region")
```

Comment:
I wonder whether we want/need some validation on this field against a list of valid regions? Currently, you can enter anything potentially with small typos that could cause issues.
Perhaps even a future goal would be to provide a list of regions that the user can select from rather than allowing free text?

Reply:
The original thinking here was that we only wanted to do some light validation and let the TF errors come through to the user; that said it would be trivial to add a simple regex test against the region format.  I think giving a list to choose from may be more maintenance than necessary.

Reply:
A regex could be the way to go on this for some light validation. Agreed on the point of maintenance.